### PR TITLE
ttl zero callback + get tasks generator

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by venthur** [comment]
Can you write a test that checks if the callback is called after the expected `ttl`?

File: redistq/task_queue.py [hunk 1]
```diff
@@ -174,6 +179,19 @@ def get(self, lease_timeout):
...
         return task['task'], task_id
 
+    def get_tasks(self, lease_timeout):
+        while True:
+            task, id_ = self.get(lease_timeout)
+            if task is not None:
+                yield task, id_
+                self.complete(id_)
+            if self.is_empty():
+                logger.info(
```
Comment (jacopofar) [inline redistq/task_queue.py]: I would remove this logging, or at least put it as debug and remove "stopping worker", it may make no sense for the generic use case

File: redistq/task_queue.py [hunk 2]
```diff
@@ -51,16 +53,20 @@ def __init__(self, host, name, timeout, reset=False):
         reset : bool
             If true, reset existing keys in the DB that have `name` as
             prefix.
-
+        ttl_zero_callback : function
```
Comment (venthur) [inline redistq/task_queue.py]: i believe the type should be called `callable` here.

File: redistq/task_queue.py [hunk 3]
```diff
@@ -174,6 +179,19 @@ def get(self, lease_timeout):
 
         return task['task'], task_id
 
+    def get_tasks(self, lease_timeout):
```
Comment (venthur) [inline redistq/task_queue.py]: why did you add this function? and if that function is necessary, shouldn't it be an iterator (i.e. implement the [iterator protocol](https://docs.python.org/3/library/stdtypes.html#iterator-types), so you can do stuff like:
```python
for task in tq:
    # do something
```

Comment (jacopofar) [reply redistq/task_queue.py]: Yes that's the intended usage. We discussed it and noticed that the user will almost always want to do that (iterate over work units until there are and marking them as done), so it makes sense to have this logic already presented as a generator.
I didn't think about implementing it using the protocol but it definitely makes sense to me

Comment (venthur) [reply redistq/task_queue.py]: That would mean that the `lease_timeout` has to move into the constructor. Otherwise, I think I'm fine with it it is just a bit strange at first that you have an iterator that can block for hours before returning the next item :D

Comment (ghost) [reply redistq/task_queue.py]: The `lease_timeout` just sets the deadline for when we want to move a job from processing to task queue. Hence it has no effect on blocking. Perhaps that also informs your point about moving to the constructor. One may want to have different lease timeouts for different `get_tasks` calls.

Comment (venthur) [reply redistq/task_queue.py]: > The lease_timeout just sets the deadline for when we want to move a job from processing to task queue. Hence it has no effect on blocking.

eventually all workers have to block until the last job finishes right? an they will block at most `least_timeout` seconds.

> Perhaps that also informs your point about moving to the constructor. One may want to have different lease timeouts for different get_tasks calls.

if you want to implement a proper iterator protocol, your `__iter__` and `__next__` methods cannot have the timeout as parameter -- so you *have* to move this config into the constructor.
Also, I can't imagine a scenario where you want different `lease_timeout`s for different `get_tasks` calls. How do you know which task you'll get? That info would be needed in order to decide on different timeouts. The core assumption here is, that all tasks are uniform in a sense that they'll take about the same amount of time to compute.

File: redistq/task_queue.py [hunk 4]
```diff
@@ -45,22 +48,30 @@ def __init__(self, host, name, timeout, reset=False):
             hostname of the redis server
         name : str
             name of the task queue
+        lease_timeout : int
```
Comment (ghost) [inline redistq/task_queue.py:51]: Moving this here hinders the reusability of the queue. You might want different `lease_timeout`s for different `get` or `get_task` calls. Not in our use-case but generally.

File: redistq/task_queue.py (hunk 3 above)
Comment (ghost) [reply redistq/task_queue.py]: > eventually all workers have to block until the last job finishes right? an they will block at most least_timeout seconds.

Yes but that's a feature. It's still the case in the latest implementation.

Ah ok I wasn't sold on implementing it as an`__iter__`.

Comment (jacopofar) [reply redistq/task_queue.py]: I wonder now whether it makes sense to have a timeout different than the lease timeout (which depends on the task and needs to be provided). The workers should expect the queue to be already available when it starts, and the responsibility of populating it in time is not of this library. We could make it simpler by completely removing this timeout

Comment (venthur) [reply redistq/task_queue.py]: I agree, making sure that redis is ready should be ensured some other way. docker supports a health check, i assume kubernetes does the smame.

See the `.github/workflows/python-package.yaml` file for an example for docker.

File: redistq/task_queue.py [hunk 5]
```diff
@@ -167,13 +171,44 @@ def get(self, lease_timeout):
...
         now = time.time()
-        deadline = now + lease_timeout
+        deadline = now + self.lease_timeout
         task = self._deserialize(self.conn.get(self._tasks + task_id))
         task['deadline'] = deadline
         self.conn.set(self._tasks + task_id, self._serialize(task))
 
         return task['task'], task_id
 
+    def __iter__(self):
```
Comment (venthur) [inline redistq/task_queue.py:183]: you're still missing the `__next__` and `StopIteration` error that make it a proper [iterator](https://docs.python.org/3/library/stdtypes.html#typeiter)

File: redistq/task_queue.py (hunk 4 above)
Comment (venthur) [reply redistq/task_queue.py:51]: I still don't get this point. Before, we always provided the `lease_timeout` on the `get` call which was then used to set the deadline for the task you're about to get. However, you **never** knew in advance which task you'd get, so you have to assume the same `lease_timeout` for all of them otherwise you'd end up with random `lease_timeout`s for tasks. Am I missing something?
